requests_kerberos/kerberos_.py

```diff
@@ -161,14 +162,44 @@ def _get_channel_bindings_application_data(response):
 
     return application_data
 
+
+# Return the hostname as it will be canonicalized by
+# krb5_sname_to_principal.  We can't simply use socket.getfqdn()
+# because it explicitly prefers results containing periods and
+# krb5_sname_to_principal doesn't care.
+#
+# This is necessary to support use cases where multiple services
+# are hosted on the same machine under multiple CNAMEs. It won't do
+# any good to use a SPN that mentions one of the CNAMES, we need to ask for the real machine.
+#
+# Maybe we should use socket.getfqdn here, but that doesn't use AI_CANONNAME
+# but rather chooses the first entry in gethostbyaddr that contains a dot, which seems
+# a trifle strange: https://github.com/python/cpython/blob/master/Lib/socket.py#L681
+#
+# https://github.com/krb5/krb5/blob/d406afa363554097ac48646a29249c04f498c88e/src/util/k5test.py#L505-L520
+# https://chromium.googlesource.com/chromium/src/+/lkgr/net/http/http_auth_handler_negotiate.cc#150
+def _get_default_kerb_host(hostname):
+    try:
+        ais = socket.getaddrinfo(hostname, None, 0, 0, 0, socket.AI_CANONNAME)
+    except socket.gaierror:
+        ais = []
+    
+    if not ais:
+        return hostname.lower()
+
+    _family, _socktype, _proto, canonname, _sockaddr = ais[0]
+    return canonname.lower()
+
+
 class HTTPKerberosAuth(AuthBase):
     """Attaches HTTP GSSAPI/Kerberos Authentication to the given Request
     object."""
     def __init__(
             self, mutual_authentication=REQUIRED,
             service="HTTP", delegate=False, force_preemptive=False,
             principal=None, hostname_override=None,
-            sanitize_mutual_error_response=True, send_cbt=True):
+            sanitize_mutual_error_response=True, send_cbt=True,
+            canonicalize_hostname=False):
```

Comment:
I can foresee someone asking us to also reverse canonicalize this in the future. I'm not saying you should implement that, but naming the argument canonicalize_hostname and having it a boolean can make the API messy in the future if we do add that.

Maybe we should make it an enumerable of sorts, like mutual_authentication. The default can switch from False to None, and we can declare a module level constant like REQUIRED/OPTIONAL/DISABLED, but named FORWARD. That leaves us the option to later make another option named REVERSE.